ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -17,6 +17,121 @@
+         deserialized snapshot to create a @ExtLedgerState blk ValuesMK@ which
+         can be 1. used in the legacy database and 2. after stripping out the
+         @ValuesMK@, used in the modern database.
+
+       - __Run both__ + snapshot doesn't have UTxO inside: an error will be
+         thrown as we cannot/don't think we should reconstruct an in-memory UTxO
+         directly from the @BackingStore@.
+
+       - Otherwise: Strip out the values inside the @ExtLedgerState blk EmptyMK@
+         to make sure that we don't have an in-memory UTxO map and use it as the
+         anchor for the modern database.
+
+  4. Replay on top of this ledger database all blocks up to the immutable
+     database tip.
+
+  At this point, the node carries a @LedgerDB@ that is initialized and ready to
+  be applied blocks on the volatile database.
+
+  == Taking snapshots during normal operation
+
+  Snapshots are taken by the @'copyAndSnapshotRunner'@ when the disk policy
+  dictates to do so. Whenever the chain grows past @k@ blocks, said runner will
+  copy the blocks which are more than @k@ blocks from the tip (i.e. the ones
+  that must be considered immutable) to the immutable database and then:
+
+  1. Flush differences in the modern DB up to the immutable db tip.
+
+  2. If dictated by the disk policy, serialize an @ExtLedgerState blk EmptyMK@ containing:
+
+       - If the legacy database is enabled, the state to serialize is the anchor
+         of the legacy database (which should be the same as the one from the
+         modern database) with the tables stowed.
+
+       - Otherwise, the state to serialize is the modern database anchor which
+         requires no further modifications.
+
+  3. A directory is created named after the slot number of the ledger state
+     being snapshotted, and the serialization from the previous step is written
+     into the @\<slotNumber\>/state@ file and the @BackingStore@ tables are copied
+     into the @\<slotNumber\>/tables@ file.
+
+  4. There is a maximum number of snapshots that should exist in the disk at any
+     time, dictated by the @DiskPolicy@, so if needed, we will trim out old
+     snapshots.
+
+  == Taking snapshots at startup
+
+  Due to the nature of the modern database having to carry around all the
+  differences between the last snapshotted state and the current tip, there is a
+  need to flush when replaying the chain as otherwise, for example on a replay
+  from genesis to the tip, we would carry millions of differences in memory.
+
+  Because of this, when we are replaying blocks we will flush regularly. As the
+  last snapshot that was taken lives in a @\<slotNumber\>/tables@ file, there is
+  no risk of destroying it by flushing. Only when we finish replaying blocks and
+  start the background threads (and specifically the @copyAndSnapshotRunner@),
+  we will take a snapshot of the current immutable database anchor as described
+  above.
```

Comment:
Is it possible that in the future we take snapshots while replaying?

Reply:
Actually, I wonder why not implement this behaviour now :thinking:

Reply:
Could be, don't see why not